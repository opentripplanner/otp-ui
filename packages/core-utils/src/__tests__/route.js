import {
  getMostReadableTextColor,
  getTransitOperatorFromLeg,
  getTransitOperatorFromOtpRoute,
  makeRouteComparator
} from "../route";

const { otp1Routes, otp2Routes } = require("./__mocks__/routes.json");

function sortRoutes(...routes) {
  routes.sort(makeRouteComparator());
  return routes;
}

describe("util > route", () => {
  describe("getTransitOperatorFromLeg", () => {
    it("should get a transit operator", () => {
      const expectedTransitOperator = { agencyId: "abc", feedId: "1" };
      expect(
        getTransitOperatorFromLeg({ agencyId: "abc", routeId: "1:rr" }, [
          { agencyId: "abc", feedId: "2" },
          expectedTransitOperator
        ])
      ).toBe(expectedTransitOperator);
    });

    it("should return null if transit operator is not found", () => {
      expect(getTransitOperatorFromLeg("abc", [])).toBeNull();
    });
  });

  describe("getTransitOperatorFromOtp1Route", () => {
    const otpRoute = { agencyId: "abc", id: "1:abc" };
    it("should get a transit operator", () => {
      const expectedTransitOperator = { agencyId: "abc", feedId: "1" };
      expect(
        getTransitOperatorFromOtpRoute(otpRoute, [
          { agencyId: "abc", feedId: "2" },
          expectedTransitOperator
        ])
      ).toBe(expectedTransitOperator);
    });

    it("should return null if transit operator is not found", () => {
      expect(getTransitOperatorFromOtpRoute(otpRoute, [])).toBeNull();
    });
  });

  describe("routeComparator", () => {
    it("should sort routes based off of agencyName if no transitOperators are defined", () => {
      expect(sortRoutes(otp1Routes[15], otp1Routes[16])).toMatchSnapshot();
    });

    it("should sort routes based off of transitOperators sort value when a match is found", () => {
      const comparatorWithTransitOperators = makeRouteComparator([
        { agencyId: "abc", feedId: "1", order: 2 },
        { agencyId: "abc", feedId: "2", order: 1 }
      ]);
      const routes = [otp1Routes[15], otp1Routes[16]];
      expect(routes.sort(comparatorWithTransitOperators)).toMatchSnapshot();
    });

    it("should sort routes based off of sortOrder", () => {
      expect(sortRoutes(otp1Routes[0], otp1Routes[1])).toMatchSnapshot();
    });

    it("should prioritize OTP1 routes with valid sortOrder", () => {
      expect(sortRoutes(otp1Routes[1], otp1Routes[2])).toMatchSnapshot();
    });

    it("should prioritize OTP2 routes with valid sortOrder", () => {
      expect(
        sortRoutes(otp2Routes[0], otp2Routes[2], otp2Routes[3])
      ).toMatchSnapshot();
    });

    it("should sort routes based off of integer shortName", () => {
      expect(sortRoutes(otp1Routes[2], otp1Routes[3])).toMatchSnapshot();
    });

    it("should prioritize routes with alphabetic shortNames over integer shortNames", () => {
      expect(sortRoutes(otp1Routes[3], otp1Routes[4])).toMatchSnapshot();
    });

    it("should sort routes based off of shortNames", () => {
      expect(sortRoutes(otp1Routes[4], otp1Routes[5])).toMatchSnapshot();
    });

    it("should sort routes with alphanumeric shortNames", () => {
      expect(sortRoutes(otp1Routes[13], otp1Routes[14])).toMatchSnapshot();
    });

    it("should prioritize routes with shortNames over those with just longNames", () => {
      expect(sortRoutes(otp1Routes[5], otp1Routes[6])).toMatchSnapshot();
    });

    it("should sort routes based off of longNames", () => {
      expect(sortRoutes(otp1Routes[8], otp1Routes[9])).toMatchSnapshot();
    });
    it("should sort OTP1 routes on all of the criteria at once", () => {
      expect(sortRoutes(...otp1Routes)).toMatchSnapshot();
    });
    it("should sort OTP2 routes on all of the criteria at once", () => {
      expect(sortRoutes(...otp2Routes)).toMatchSnapshot();
    });

    it("should sort based off of route type", () => {
      expect(sortRoutes(otp1Routes[11], otp1Routes[12])).toMatchSnapshot();
    });
  });

  // These color pairings are taken from real agencies.
  // See the method jsdoc for more details.
  describe("route text color generation", () => {
    it("should not modify appropriate color pairings", () => {
      expect(getMostReadableTextColor("#EE352E", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#00933C", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#b933ad", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#808183", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#0039a6", "#ffffff")).toBe("#ffffff");

      // Both colors are readable and used by transit agencies.
      // The function accepts both colors.
      expect(getMostReadableTextColor("#ff6319", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#ff6319", "#000000")).toBe("#000000");
      expect(getMostReadableTextColor("#6cbe45", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#6cbe45", "#000000")).toBe("#000000");
      expect(getMostReadableTextColor("#a7a9ac", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#a7a9ac", "#000000")).toBe("#000000");
      expect(getMostReadableTextColor("#FF00FF", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#FF00FF", "#000000")).toBe("#000000");
      expect(getMostReadableTextColor("#009999", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#009999", "#000000")).toBe("#000000");

      expect(getMostReadableTextColor("#996633", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#fccc0a", "#000000")).toBe("#000000");

      expect(getMostReadableTextColor("#0075B2", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#D4A723", "#000000")).toBe("#000000");
      expect(getMostReadableTextColor("#009D4B", "#ffffff")).toBe("#ffffff");
      expect(getMostReadableTextColor("#CE242B", "#ffffff")).toBe("#ffffff");

      expect(getMostReadableTextColor("#00FF00", "#000000")).toBe("#000000");

      expect(getMostReadableTextColor("#E32400", "#FFFFFF")).toBe("#FFFFFF");
      expect(getMostReadableTextColor("#d18881", "#000000")).toBe("#000000");

      expect(getMostReadableTextColor("#004080", "#FFFFFF")).toBe("#FFFFFF");
    });

    it("should modify inappropriate color pairings", () => {
      expect(getMostReadableTextColor("#EE352E", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#00933C", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#b933ad", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#808183", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#0039a6", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#996633", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#fccc0a", "#ffffff")).toBe("#000000");

      expect(getMostReadableTextColor("#0075B2", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#D4A723", "#ffffff")).toBe("#000000");
      expect(getMostReadableTextColor("#009D4B", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#CE242B", "#000000")).toBe("#ffffff");

      expect(getMostReadableTextColor("#00FF00", "#ffffff")).toBe("#000000");
      expect(getMostReadableTextColor("#321b26", "#000000")).toBe("#ffffff");
      expect(getMostReadableTextColor("#321b26", "9e4141")).toBe("#ffffff");

      expect(getMostReadableTextColor("#f00", "#f00")).toBe("#ffffff");
      expect(getMostReadableTextColor("#fff", "#fff")).toBe("#000000");
      expect(getMostReadableTextColor("#000", "#000")).toBe("#ffffff");
    });
  });
});
