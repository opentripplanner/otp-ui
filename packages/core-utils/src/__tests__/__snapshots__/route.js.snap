// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util > route routeComparator should prioritize OTP1 routes with valid sortOrder 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Around town",
    "mode": "BUS",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
    "type": 3,
  },
]
`;

exports[`util > route routeComparator should prioritize OTP2 routes with valid sortOrder 1`] = `
Array [
  Object {
    "longName": "Loop route",
    "mode": "BUS",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "longName": "Across town",
    "mode": "BUS",
    "shortName": "10",
    "sortOrder": 10,
  },
  Object {
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": null,
    "type": 3,
  },
]
`;

exports[`util > route routeComparator should prioritize routes with alphabetic shortNames over integer shortNames 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "A-line",
    "mode": "BUS",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Loop route",
    "mode": "BUS",
    "shortName": "2",
    "sortOrder": -999,
  },
]
`;

exports[`util > route routeComparator should prioritize routes with shortNames over those with just longNames 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "B-line",
    "mode": "BUS",
    "shortName": "B",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "A meandering route",
    "mode": "BUS",
    "sortOrder": -999,
  },
]
`;

exports[`util > route routeComparator should sort OTP1 routes on all of the criteria at once 1`] = `
Array [
  Object {
    "agencyId": "abc",
    "agencyName": "Acclaimed agency 1",
    "longName": "Express route X",
    "mode": "BUS",
    "routeId": "2:erx",
    "shortName": "30X",
    "sortOrder": 10,
  },
  Object {
    "agencyId": "abc",
    "agencyName": "Agency 1",
    "longName": "Intercity Train 2",
    "mode": "RAIL",
    "routeId": "1:it2",
    "shortName": "IC",
    "sortOrder": 10,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Yellow line Subway",
    "mode": "SUBWAY",
    "shortName": "Yellow",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Intercity Train",
    "mode": "RAIL",
    "shortName": "IC",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Local route",
    "mode": "BUS",
    "shortName": "6",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Around town",
    "mode": "BUS",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Variation of express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Zig-zagging route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Xpress route C",
    "mode": "BUS",
    "shortName": "30C",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Across town",
    "mode": "BUS",
    "shortName": "10",
    "sortOrder": 10,
  },
  Object {
    "agencyId": "bcd",
    "longName": "A-line",
    "mode": "BUS",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "B-line",
    "mode": "BUS",
    "shortName": "B",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Loop route",
    "mode": "BUS",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
    "type": 3,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Express route X",
    "mode": "BUS",
    "shortName": "30X",
  },
  Object {
    "agencyId": "bcd",
    "longName": "A meandering route",
    "mode": "BUS",
    "sortOrder": -999,
  },
]
`;

exports[`util > route routeComparator should sort OTP2 routes on all of the criteria at once 1`] = `
Array [
  Object {
    "agency": Object {
      "id": "abc",
      "name": "Acclaimed agency 1",
    },
    "longName": "Express route X",
    "mode": "BUS",
    "routeId": "2:erx",
    "shortName": "30X",
    "sortOrder": 2,
  },
  Object {
    "agency": Object {
      "id": "abc",
      "name": "Agency 1",
    },
    "longName": "Intercity Train 2",
    "mode": "RAIL",
    "routeId": "1:it2",
    "shortName": "IC",
    "sortOrder": 2,
  },
  Object {
    "longName": "A-line",
    "mode": "BUS",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "longName": "Loop route",
    "mode": "BUS",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "longName": "Yellow line Subway",
    "mode": "SUBWAY",
    "shortName": "Yellow",
    "sortOrder": 2,
  },
  Object {
    "longName": "Intercity Train",
    "mode": "RAIL",
    "shortName": "IC",
    "sortOrder": 2,
  },
  Object {
    "longName": "Local route",
    "mode": "BUS",
    "shortName": "6",
    "sortOrder": 2,
  },
  Object {
    "longName": "Around town",
    "mode": "BUS",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "longName": "Express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Variation of express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Zig-zagging route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Xpress route C",
    "mode": "BUS",
    "shortName": "30C",
    "sortOrder": 2,
  },
  Object {
    "longName": "Express route X",
    "mode": "BUS",
    "shortName": "30X",
    "sortOrder": 2,
  },
  Object {
    "longName": "Across town",
    "mode": "BUS",
    "shortName": "10",
    "sortOrder": 10,
  },
  Object {
    "longName": "B-line",
    "mode": "BUS",
    "shortName": "B",
  },
  Object {
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": null,
    "type": 3,
  },
  Object {
    "longName": "A meandering route",
    "mode": "BUS",
  },
]
`;

exports[`util > route routeComparator should sort based off of route type 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Yellow line Subway",
    "mode": "SUBWAY",
    "shortName": "Yellow",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Intercity Train",
    "mode": "RAIL",
    "shortName": "IC",
    "sortOrder": 2,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of agencyName if no transitOperators are defined 1`] = `
Array [
  Object {
    "agencyId": "abc",
    "agencyName": "Acclaimed agency 1",
    "longName": "Express route X",
    "mode": "BUS",
    "routeId": "2:erx",
    "shortName": "30X",
    "sortOrder": 10,
  },
  Object {
    "agencyId": "abc",
    "agencyName": "Agency 1",
    "longName": "Intercity Train 2",
    "mode": "RAIL",
    "routeId": "1:it2",
    "shortName": "IC",
    "sortOrder": 10,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of integer shortName 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Loop route",
    "mode": "BUS",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
    "type": 3,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of longNames 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Variation of express route",
    "mode": "BUS",
    "shortName": "30",
    "sortOrder": 2,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of shortNames 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "A-line",
    "mode": "BUS",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "agencyId": "bcd",
    "longName": "B-line",
    "mode": "BUS",
    "shortName": "B",
    "sortOrder": -999,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of sortOrder 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Around town",
    "mode": "BUS",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Across town",
    "mode": "BUS",
    "shortName": "10",
    "sortOrder": 10,
  },
]
`;

exports[`util > route routeComparator should sort routes based off of transitOperators sort value when a match is found 1`] = `
Array [
  Object {
    "agencyId": "abc",
    "agencyName": "Agency 1",
    "longName": "Intercity Train 2",
    "mode": "RAIL",
    "routeId": "1:it2",
    "shortName": "IC",
    "sortOrder": 10,
  },
  Object {
    "agencyId": "abc",
    "agencyName": "Acclaimed agency 1",
    "longName": "Express route X",
    "mode": "BUS",
    "routeId": "2:erx",
    "shortName": "30X",
    "sortOrder": 10,
  },
]
`;

exports[`util > route routeComparator should sort routes with alphanumeric shortNames 1`] = `
Array [
  Object {
    "agencyId": "bcd",
    "longName": "Xpress route C",
    "mode": "BUS",
    "shortName": "30C",
    "sortOrder": 2,
  },
  Object {
    "agencyId": "bcd",
    "longName": "Express route X",
    "mode": "BUS",
    "shortName": "30X",
  },
]
`;
